### PR TITLE
refactor(bits-array): fix potential deadlock in bit_array.go

### PR DESCRIPTION
**Pull request by shotonoff**
## Issue being fixed or feature implemented

Detected potential deadlock in an implementation of bit-array component

## What was done?

Refactored `bit-array` component to eliminate deadlock issue and make a code more unambiguous 

## How Has This Been Tested?

Unit-test

## Breaking Changes

N/A

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: libs/bits/bit_array.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ import (
 
 // BitArray is a thread-safe implementation of a bit array.
 type BitArray struct {
-	mtx   tmsync.Mutex
+	mtx   tmsync.RWMutex
```
Comment (lklimek) [inline libs/bits/bit_array.go:20]: Please also change to RLock in places like getIndex

File: libs/bits/bit_array.go [hunk 2]
```diff
@@ -126,15 +126,14 @@ func (bA *BitArray) Or(o *BitArray) *BitArray {
 	if o == nil {
 		return bA.Copy()
 	}
+	o = o.Copy()
 	bA.mtx.Lock()
-	o.mtx.Lock()
+	defer bA.mtx.Unlock()
```
Comment (lklimek) [inline libs/bits/bit_array.go:131]: Maybe use RLock / RUnlock here?

File: libs/bits/bit_array.go [hunk 3]
```diff
@@ -126,15 +126,14 @@ func (bA *BitArray) Or(o *BitArray) *BitArray {
 	if o == nil {
 		return bA.Copy()
 	}
+	o = o.Copy()
```
Comment (lklimek) [inline libs/bits/bit_array.go:129]: Maybe this should be AFTER bA.mtx.RLock, to ensure that neither bA nor o changed during execution of this function

File: libs/bits/bit_array.go [hunk 4]
```diff
@@ -145,12 +144,9 @@ func (bA *BitArray) And(o *BitArray) *BitArray {
 	if bA == nil || o == nil {
 		return nil
 	}
+	o = o.Copy()
```
Comment (lklimek) [inline libs/bits/bit_array.go:147]: Maybe this should be AFTER bA.mtx.RLock, to ensure that neither bA nor o changed during execution of this function

File: libs/bits/bit_array.go [hunk 5]
```diff
@@ -145,12 +144,9 @@ func (bA *BitArray) And(o *BitArray) *BitArray {
...
 	bA.mtx.Lock()
-	o.mtx.Lock()
-	defer func() {
-		bA.mtx.Unlock()
-		o.mtx.Unlock()
-	}()
+	defer bA.mtx.Unlock()
```
Comment (lklimek) [inline libs/bits/bit_array.go:149]: Maybe use RLock / RUnlock here?

(Also, in other similar places below)

File: libs/bits/bit_array.go (hunk 3 above)
Comment (shotonoff) [reply libs/bits/bit_array.go:129]: if to add a condition to check `o` and `bA` on equality, anyway go-deadlock reports about potential deadlock issue

File: libs/bits/bit_array.go (hunk 4 above)
Comment (shotonoff) [reply libs/bits/bit_array.go:147]: answered above

**Review by shotonoff — COMMENTED** [review]
please check answers

**Review by lklimek — APPROVED** [review]

**Pull request merged**